diff --git a/src/core/ipv4/dhcp.c b/src/core/ipv4/dhcp.c
index b519ae909..96de64423 100644
--- a/src/core/ipv4/dhcp.c
+++ b/src/core/ipv4/dhcp.c
@@ -703,6 +703,10 @@ dhcp_handle_ack(struct netif *netif, struct dhcp_msg *msg_in)
     dns_setserver(n, &dns_addr);
   }
 #endif /* LWIP_DHCP_PROVIDE_DNS_SERVERS */
+  ip4_addr_set_zero(&dhcp->offered_dns_addr);
+  if (dhcp_option_given(dhcp, DHCP_OPTION_IDX_DNS_SERVER)) {
+    ip4_addr_set_u32(&dhcp->offered_dns_addr, lwip_htonl(dhcp_get_option_value(dhcp, DHCP_OPTION_IDX_DNS_SERVER)));
+  }
 }
 
 /**
diff --git a/src/include/lwip/dhcp.h b/src/include/lwip/dhcp.h
index 962cc2ee2..752db1f67 100644
--- a/src/include/lwip/dhcp.h
+++ b/src/include/lwip/dhcp.h
@@ -103,6 +103,7 @@ struct dhcp
   ip4_addr_t offered_ip_addr;
   ip4_addr_t offered_sn_mask;
   ip4_addr_t offered_gw_addr;
+  ip4_addr_t offered_dns_addr;
 
   u32_t offered_t0_lease; /* lease period (in seconds) */
   u32_t offered_t1_renew; /* recommended renew time (usually 50% of lease period) */
diff --git a/src/include/netif/ppp/ppp.h b/src/include/netif/ppp/ppp.h
index 3d73c3657..2f1e1e9e3 100644
--- a/src/include/netif/ppp/ppp.h
+++ b/src/include/netif/ppp/ppp.h
@@ -421,6 +421,8 @@ struct ppp_pcb_s {
   ipv6cp_options ipv6cp_allowoptions; /* Options we allow peer to request */
   ipv6cp_options ipv6cp_hisoptions;   /* Options that we ack'd */
 #endif /* PPP_IPV6_SUPPORT */
+
+  u32_t dns_server;
 };
 
 /************************
diff --git a/src/netif/ppp/ipcp.c b/src/netif/ppp/ipcp.c
index b7c766eb0..2d33f97c2 100644
--- a/src/netif/ppp/ipcp.c
+++ b/src/netif/ppp/ipcp.c
@@ -1930,7 +1930,9 @@ static void ipcp_up(fsm *f) {
     if (go->dnsaddr[1])
 	script_setenv("DNS2", ip_ntoa(go->dnsaddr[1]), 0);
 #endif /* UNUSED */
+    pcb->dns_server = 0;
     if (pcb->settings.usepeerdns && (go->dnsaddr[0] || go->dnsaddr[1])) {
+	pcb->dns_server = go->dnsaddr[0];
 	sdns(pcb, go->dnsaddr[0], go->dnsaddr[1]);
 #if 0 /* UNUSED */
 	script_setenv("USEPEERDNS", "1", 0);
